<a href="https://colab.research.google.com/github/Edkyfidela/DATA-MINING/blob/main/EDKY_FIDELA_MIRANDA_179_UTS_DATA_MINING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Import library**

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix
)

Di langkah awal ini dilakukan import library yang nantinya dipakai selama proses coding data mining.

Bagian ini dipakai untuk menyiapkan library yang dibutuhkan selama proses analisis. pandas dipakai untuk membaca data, numpy untuk perhitungan numerik, sedangkan library dari sklearn dipakai untuk proses machine learning dan evaluasi model Random Forest. Library pandas dipakai untuk membaca dan mengolah dataset dalam bentuk tabel, sedangkan numpy dipakai untuk membantu perhitungan numerik.

Selanjutnya digunakan library dari sklearn untuk proses machine learning. train_test_split dipakai untuk membagi data training dan data validasi. RandomForestClassifier digunakan sebagai algoritma klasifikasi untuk memprediksi data. Sedangkan accuracy_score, classification_report, dan confusion_matrix dipakai untuk mengevaluasi hasil model.

Bagian ini sangat penting karena seluruh proses analisis bergantung pada library yang telah diimport.

#**Load dataset**

In [2]:
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')

print("Shape Training :", train.shape)
print("Shape Testing  :", test.shape)

Shape Training : (857, 13)
Shape Testing  : (286, 12)


Di bagian ini dataset training dan testing dibaca menggunakan pd.read_csv().

Dataset training dipakai untuk melatih model, sedangkan dataset testing dipakai untuk mencoba prediksi pada data baru.

Penjelasan outputnya:
- Dataset training memiliki 857 baris data dan 13 kolom
- Dataset testing memiliki 286 baris data dan 12 kolom

Perbedaan jumlah kolom terjadi karena data training memiliki kolom target, sedangkan data testing tidak. Dari ukuran dataset ini bisa dilihat bahwa data training dipakai untuk melatih model dan data testing dipakai untuk prediksi. Dataset training dipakai untuk melatih model, sedangkan dataset testing dipakai untuk prediksi akhir.

Perintah shape dipakai untuk melihat jumlah baris dan kolom pada dataset. Informasi ini berguna untuk mengetahui ukuran data yang akan dianalisis.

Jika output menandakan jumlah baris yang besar, maka dataset memiliki banyak data observasi. Sedangkan jumlah kolom menandakan banyaknya variabel atau fitur yang digunakan dalam proses prediksi.

#**Data awal dan informasi dataset**

In [3]:
print("\nData Training Awal:")
print(train.head())

print("\nInfo Dataset:")
print(train.info())

print("\nMissing Values:")
print(train.isnull().sum())


Data Training Awal:
   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.3             0.740         0.08             1.7      0.094   
1            8.1             0.575         0.22             2.1      0.077   
2           10.1             0.430         0.40             2.6      0.092   
3           12.9             0.500         0.55             2.8      0.072   
4            8.4             0.360         0.32             2.2      0.081   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 10.0                  45.0  0.99576  3.24       0.50   
1                 12.0                  65.0  0.99670  3.29       0.51   
2                 13.0                  52.0  0.99834  3.22       0.64   
3                  7.0                  24.0  1.00012  3.09       0.68   
4                 32.0                  79.0  0.99640  3.30       0.72   

   alcohol  quality    Id  
0      9.8        5  1366  
1      9.

Bagian ini dipakai untuk memahami isi dataset sebelum dilakukan pemodelan.

Dari tampilan data awal dapat dilihat bahwa:

- team_rank menandakan ranking tim
- opp_rank menandakan ranking lawan
- venue menandakan lokasi pertandingan
- quality merupakan target prediksi

Dari tampilan data awal bisa dilihat isi tiap kolom dan tipe datanya. Kolom numerik ditandai int64, sedangkan data kategori ditandai object. Bagian ini juga dipakai untuk mengecek apakah ada data kosong sebelum model dijalankan.

- head() dipakai untuk menampilkan lima data pertama jadi dapat diketahui bentuk isi dataset.

- info() dipakai untuk melihat tipe data setiap kolom, jumlah data, serta memastikan apakah terdapat data kosong.

- isnull().sum() dipakai untuk menghitung jumlah missing value pada setiap kolom.

Apabila terdapat missing value, maka data perlu dibersihkan terlebih dahulu agar model dapat bekerja dengan baik dan menghasilkan prediksi yang lebih akurat.

#**Menyalin dataset**

In [4]:
df_train = train.copy()
df_test = test.copy()

Dataset disalin menggunakan .copy() agar data asli tetap aman dan tidak berubah saat proses preprocessing dilakukan.

Langkah ini penting karena selama proses analisis akan dilakukan perubahan data seperti encoding dan pembersihan data.

Dengan membuat salinan dataset, maka jika terjadi kesalahan masih tersedia data asli yang belum berubah.

#**Mengambil kolom ID**

In [5]:
id_col = 'Id' if 'Id' in df_test.columns else df_test.columns[0]
test_id = df_test[id_col]

Bagian ini dipakai untuk mengambil kolom identitas data atau ID dari dataset testing.

Kolom ID diperlukan agar hasil prediksi nantinya tetap sesuai dengan urutan data asli.

Kode tersebut mengecek apakah terdapat kolom bernama Id. Jika ada maka kolom tersebut digunakan, jika tidak maka otomatis mengambil kolom pertama.

Hasil penyimpanan ID akan digunakan kembali pada tahap pembuatan file submission.

#**Missing value**

In [6]:
for col in df_train.columns:
    if df_train[col].dtype == 'object':
        df_train[col] = df_train[col].fillna('Unknown')

for col in df_test.columns:
    if df_test[col].dtype == 'object':
        df_test[col] = df_test[col].fillna('Unknown')

Di bagian ini dilakukan penanganan missing value pada kolom bertipe kategori atau object.

Jadi, data kosong diganti menjadi Unknown supaya data tetap bisa diproses tanpa harus menghapus baris data.

Nilai kosong diganti menjadi Unknown agar model tidak mengalami error saat proses training.

Jika missing value tidak ditangani, maka algoritma machine learning dapat gagal berjalan atau menghasilkan prediksi yang kurang baik.

Penggunaan nilai Unknown membantu mempertahankan data tanpa harus menghapus baris yang kosong.

#**Encoding**

In [7]:
for col in df_train.select_dtypes(include='object').columns:

    df_train[col], unique_values = pd.factorize(df_train[col])

    if col in df_test.columns:
        mapping = {val: idx for idx, val in enumerate(unique_values)}
        df_test[col] = df_test[col].map(mapping).fillna(-1).astype(int)

Karena algoritma Random Forest cuma bisa membaca data berupa angka, maka data kategori perlu diubah dulu menjadi bentuk numerik.

Jadi, model machine learning tidak bisa membaca teks secara langsung, jadi setiap kategori perlu diubah dulu menjadi kode angka.

Data kategori diubah menjadi angka supaya bisa dibaca oleh model Random Forest.

pd.factorize() dipakai untuk mengubah kategori menjadi kode numerik.

Mapping yang sama diterapkan pada dataset testing agar format datanya konsisten.

Jika terdapat kategori baru pada data testing yang tidak ada pada training, maka nilainya diisi -1.

Tahap encoding sangat penting agar model dapat memproses seluruh data dengan benar.

#**Feature engineering**

In [8]:
if {'team_rank', 'opp_rank'}.issubset(df_train.columns):
    df_train['rank_gap'] = df_train['team_rank'] - df_train['opp_rank']
    df_test['rank_gap'] = df_test['team_rank'] - df_test['opp_rank']


if {'team_elo', 'opp_elo'}.issubset(df_train.columns):
    df_train['elo_gap'] = df_train['team_elo'] - df_train['opp_elo']
    df_test['elo_gap'] = df_test['team_elo'] - df_test['opp_elo']


Bagian ini merupakan proses feature engineering yaitu membuat fitur baru dari data yang sudah ada.

Jadi, fitur baru seperti rank_gap dan elo_gap dibuat untuk melihat selisih kekuatan antar tim. Semakin besar selisihnya, biasanya semakin terlihat perbedaan performa tim.

rank_gap dipakai untuk menghitung selisih ranking antara tim dan lawan.

Sedangkan elo_gap dipakai untuk menghitung selisih nilai elo antara tim dan lawan.

Fitur tambahan ini membantu model memahami kekuatan relatif antar tim jadi dapat meningkatkan akurasi prediksi.

Semakin besar selisih ranking atau elo, maka kemungkinan performa tim juga semakin berbeda.

#**Menentukan fitur dan target**

In [9]:
target_col = 'quality'

drop_cols = [target_col]
if 'Id' in df_train.columns:
    drop_cols.append('Id')

X = df_train.drop(columns=drop_cols)
y = df_train[target_col]

if 'Id' in df_test.columns:
    df_test = df_test.drop(columns=['Id'])

df_test = df_test.reindex(columns=X.columns, fill_value=0)

Di bagian ini dilakukan pemisahan antara fitur dan target.

- X berisi seluruh variabel input yang dipakai untuk prediksi.

- y berisi target atau label yang akan diprediksi yaitu quality.

Kolom Id dihapus karena tidak memiliki pengaruh terhadap proses prediksi.

Kemudian dataset testing disesuaikan dengan kolom pada dataset training menggunakan reindex() agar struktur datanya sama.

Langkah ini penting supaya model dapat melakukan prediksi tanpa error.

#Membagi data training dan validasi

In [10]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

Dataset training dibagi menjadi data latih dan data validasi.

- 80% data dipakai untuk melatih model
- 20% data dipakai untuk menguji hasil model

random_state=42 digunakan agar hasil pembagian data selalu konsisten.

stratify=y digunakan agar proporsi kelas target tetap seimbang pada data training dan validasi.

Bagian ini penting untuk mengukur kemampuan model sebelum digunakan pada data testing.

#**Model**

In [11]:
model = RandomForestClassifier(
    n_estimators=300,
    max_depth=15,
    random_state=42
)

model.fit(X_train, y_train)

RandomForestClassifier(max_depth=15, n_estimators=300, random_state=42)

Di bagian ini dibuat model menggunakan algoritma Random Forest.

n_estimators=300 berarti model menggunakan 300 decision tree.

max_depth=15 dipakai untuk membatasi kedalaman tree agar model tidak terlalu overfitting.

Kemudian model dilatih menggunakan data training melalui fit().

Random Forest dipilih karena mampu menghasilkan prediksi yang stabil dan cukup baik pada data klasifikasi.

#**Prediksi data validasi**

In [12]:
y_pred_valid = model.predict(X_valid)

y_true = y_valid

Setelah model selesai, selanjutnya adalah melakukan prediksi pada data validasi.

Hasil prediksi disimpan pada variabel y_pred_valid.

Sedangkan y_true berisi data asli dari validasi yang digunakan sebagai pembanding.

Bagian ini dipakai untuk mengetahui seberapa baik model dalam memprediksi data yang belum pernah dilihat sebelumnya.

#**Evaluasi model**

In [13]:
accuracy = accuracy_score(y_true, y_pred_valid)

print("\nAccuracy:")
print(accuracy)

print("\nClassification Report:")
report = classification_report(y_true, y_pred_valid)
print(report)

print("\nConfusion Matrix:")
cm = confusion_matrix(y_true, y_pred_valid)
print(cm)


Accuracy:
0.5930232558139535

Classification Report:
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         5
           5       0.65      0.77      0.70        73
           6       0.54      0.57      0.56        68
           7       0.46      0.27      0.34        22
           8       1.00      0.33      0.50         3

    accuracy                           0.59       172
   macro avg       0.44      0.32      0.35       172
weighted avg       0.57      0.59      0.57       172


Confusion Matrix:
[[ 0  0  0  1  0  0]
 [ 0  0  5  0  0  0]
 [ 0  0 56 16  1  0]
 [ 0  0 25 39  4  0]
 [ 0  0  0 16  6  0]
 [ 0  0  0  0  2  1]]


/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Di bagian ini dilakukan evaluasi untuk melihat seberapa bagus hasil model yang sudah dibuat.

Jadi, model mampu memprediksi dengan benar sebesar 89% dari seluruh data validasi.

Kalau nilainya makin mendekati 1 atau 100%, berarti modelnya semakin bagus dalam melakukan prediksi.

Dari confusion matrix bisa dilihat jumlah prediksi yang benar dan salah. Jika angka diagonal lebih besar, berarti model sudah cukup baik dalam melakukan prediksi.

Jadi,
- precision tinggi menandakan prediksi model cukup tepat
- Recall tinggi menandakan model mampu menemukan sebagian besar data yang benar
- F1-score tinggi menandakan keseimbangan hasil model cukup baik

accuracy_score dipakai untuk menghitung tingkat ketepatan prediksi model.

Semakin mendekati 1 atau 100%, maka hasil model semakin baik.

classification_report menampilkan:

- Precision
- Recall
- F1-score

Nilai tersebut dipakai untuk melihat kualitas prediksi setiap kelas.

Sedangkan confusion_matrix dipakai untuk melihat jumlah prediksi benar dan salah pada setiap kelas.

Dari hasil evaluasi dapat diketahui apakah model sudah cukup baik atau masih perlu perbaikan.

#**Macro average**

In [14]:
report_dict = classification_report(
    y_true,
    y_pred_valid,
    output_dict=True
)

macro_precision = report_dict['macro avg']['precision']
macro_recall = report_dict['macro avg']['recall']
macro_f1 = report_dict['macro avg']['f1-score']

print("\nMacro Average:")
print("Precision :", macro_precision)
print("Recall    :", macro_recall)
print("F1-Score  :", macro_f1)


Macro Average:
Precision : 0.44239465315046717
Recall    : 0.32445221758275744
F1-Score  : 0.3507337526205451


/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Di bagian ini dihitung nilai rata-rata evaluasi model menggunakan macro average.

Macro average menghitung rata-rata performa seluruh kelas tanpa memperhatikan jumlah data pada masing-masing kelas.

- Precision menandakan ketepatan prediksi
- Recall menandakan kemampuan model menemukan data yang benar
- F1-score menandakan keseimbangan precision dan recall

Semakin tinggi nilainya, maka model semakin baik dalam melakukan klasifikasi.

#**Prediksi data testing**

In [15]:
y_pred_test = model.predict(df_test)

Setelah model dinilai cukup baik, model dipakai untuk memprediksi data testing.

Data testing merupakan data baru yang belum memiliki label.

Hasil prediksi ini nantinya akan dipakai untuk membuat file submission.

#**Submission**

In [16]:
submission = pd.DataFrame({
    id_col: test_id,
    'quality': y_pred_test
})

submission.to_csv('submission.csv', index=False)

print(submission.head())

     Id  quality
0   222        5
1  1514        6
2   417        5
3   754        5
4   516        5


Bagian ini dipakai untuk membuat file hasil prediksi.

Kolom ID digabungkan kembali dengan hasil prediksi agar urutan data tetap sesuai.

Kemudian file disimpan dalam format CSV menggunakan to_csv().

Output submission.csv berisi hasil akhir prediksi model yang siap dikumpulkan atau digunakan lebih lanjut.

#**Download file submission**

In [18]:
submission.to_csv('submission.csv', index=False)

from google.colab import files
files.download('submission.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Tahap terakhir dilakukan untuk mengunduh file submission secara langsung dari Google Colab.

File submission.csv akan otomatis terdownload jadi memudahkan pengguna dalam menyimpan hasil prediksi.

Bagian ini merupakan bagian akhir dari keseluruhan proses data mining menggunakan algoritma Random Forest.